torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -129,15 +138,45 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
+def apply_pipeline_parallelism(model, world_mesh, parallel_dims, job_config: JobConfig):
+    assert (
+        parallel_dims.pp_enabled
+    ), "can't apply pipeline parallelism if it is not enabled"
+
+    if job_config.model.norm_type == "fused_rmsnorm":
+        # TODO(whc) - torch._dynamo.exc.Unsupported: Illegal getattr invocation stride in strict mode
+        # coming from ` if dy.stride(-1) != 1:` in fused_rmsnorm
+        raise NotImplementedError(
+            "fused_rmsnorm not yet compatible with Pipeline Tracer (strides error). Please use layernorm or rmsnorm."
+        )
+    pp_mesh = world_mesh["pp"]
+    stage_idx = pp_mesh.get_local_rank()
+    layers_per_rank = len(model.layers) // parallel_dims.pp
+    split_spec = {
+        f"layers.{i * layers_per_rank}": SplitPoint.BEGINNING
+        for i in range(1, parallel_dims.pp)
```

Comment:
I'm new to PP api and have a question:
If `layers_per_rank` = 5, `parallel_dims.pp` = 2, what should be the split_spec. My straightforward thought is `SplitPoint.BEGINNING` should contain `i = 1, 3, 5`, but according to the code it's just `i = 1`.

Reply:
`parallel_dims.pp` refers to the number of pipeline stages we split the model into. 
For example, if `model.layers` = 10, 10 // 2 = 5, then we put 5 layers per stage (i.e. `layers_per_rank = 5`).
Hence we make a cut at `model.layers.5`.